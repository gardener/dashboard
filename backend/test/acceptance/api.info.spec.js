//
// Copyright (c) 2019 by SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict'

const { version } = require('../../package')

module.exports = function info ({ agent, k8s, auth }) {
  /* eslint no-unused-expressions: 0 */
  const username = 'john.doe@example.org'
  const id = username
  const aud = [ 'gardener' ]

  it('should reject requests csrf protection error', async function () {
    const res = await agent
      .post('/api/info')
      .unset('x-requested-with')

    expect(res).to.have.status(403)
    expect(res).to.be.json
    expect(res.body.error).to.have.property('name').that.is.equal('Forbidden')
    expect(res.body.message).to.include('CSRF protection')
  })

  it('should reject requests without authorization header', async function () {
    const res = await agent
      .get('/api/info')

    expect(res).to.have.status(401)
    expect(res).to.be.json
    expect(res.body.error).to.have.property('name').that.is.equal('Unauthorized')
    expect(res.body.message).to.include('authorization token')
  })

  it('should reject requests with invalid signature', async function () {
    const user = auth.createUser({ id, aud }, true)
    const res = await agent
      .get('/api/info')
      .set('cookie', await user.cookie)

    expect(res).to.have.status(401)
    expect(res).to.be.json
    expect(res.body.error).to.have.property('name').that.is.equal('Unauthorized')
    expect(res.body.message).to.include('invalid signature')
  })

  it('should reject requests with invalid audience', async function () {
    const user = auth.createUser({ id, aud: [ 'invalid-audience' ] })
    const res = await agent
      .get('/api/info')
      .set('cookie', await user.cookie)

    expect(res).to.have.status(401)
    expect(res).to.be.json
    expect(res.body.error).to.have.property('name').that.is.equal('Unauthorized')
    expect(res.body.message).to.include('audience invalid')
  })

  it('should return information with version', async function () {
    const user = auth.createUser({ id, aud })
    const gardenerVersion = { major: '1', minor: '0' }
    k8s.stub.fetchGardenerVersion({ version: gardenerVersion })
    const res = await agent
      .get('/api/info')
      .set('cookie', await user.cookie)

    expect(res).to.have.status(200)
    expect(res).to.be.json
    expect(res.body).to.have.property('version').that.is.equal(version)
    expect(res.body.gardenerVersion).to.have.property('major').that.is.equal(gardenerVersion.major)
    expect(res.body.gardenerVersion).to.have.property('minor').that.is.equal(gardenerVersion.minor)
    expect(res.body).not.to.have.property('user')
  })

  it('should return information without version', async function () {
    const user = auth.createUser({ id, aud })
    k8s.stub.fetchGardenerVersion({ version: undefined })
    const res = await agent
      .get('/api/info')
      .set('cookie', await user.cookie)

    expect(res).to.have.status(200)
    expect(res).to.be.json
    expect(res.body).to.have.property('version').that.is.equal(version)
    expect(res.body).not.to.have.property('gardenerVersion')
    expect(res.body).not.to.have.property('user')
  })
}
