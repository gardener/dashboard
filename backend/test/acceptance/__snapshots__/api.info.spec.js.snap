// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api info should reject requests csrf protection error 1`] = `
Object {
  "code": 403,
  "details": Any<Object>,
  "message": "Request has been blocked by CSRF protection",
  "reason": "Forbidden",
  "status": "Failure",
}
`;

exports[`api info should reject requests with invalid audience 1`] = `
Object {
  "code": 401,
  "details": Any<Object>,
  "message": "jwt audience invalid. expected: gardener",
  "reason": "Unauthorized",
  "status": "Failure",
}
`;

exports[`api info should reject requests with invalid signature 1`] = `
Object {
  "code": 401,
  "details": Any<Object>,
  "message": "invalid signature",
  "reason": "Unauthorized",
  "status": "Failure",
}
`;

exports[`api info should reject requests without authorization header 1`] = `
Object {
  "code": 401,
  "details": Any<Object>,
  "message": "No authorization token was found",
  "reason": "Unauthorized",
  "status": "Failure",
}
`;

exports[`api info should return information with version 1`] = `
Array [
  Array [
    Object {
      ":authority": "kubernetes:6443",
      ":method": "get",
      ":path": "/apis/apiregistration.k8s.io/v1/apiservices/v1beta1.core.gardener.cloud",
      ":scheme": "https",
      "authorization": "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwic3ViIjoic3lzdGVtOnNlcnZpY2VhY2NvdW50OmdhcmRlbjpkZWZhdWx0In0.-4rSuvvj5BStN6DwnmLAaRVbgpl5iCn2hG0pcqx0NPw",
    },
  ],
  Array [
    Object {
      ":authority": "gardener-apiserver.gardener",
      ":method": "get",
      ":path": "/version",
      ":scheme": "https",
    },
  ],
]
`;

exports[`api info should return information with version 2`] = `
Object {
  "gardenerVersion": Object {
    "major": "1",
    "minor": "0",
  },
  "version": "1.47.0",
}
`;

exports[`api info should return information without version 1`] = `
Array [
  Array [
    Object {
      ":authority": "kubernetes:6443",
      ":method": "get",
      ":path": "/apis/apiregistration.k8s.io/v1/apiservices/v1beta1.core.gardener.cloud",
      ":scheme": "https",
      "authorization": "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwic3ViIjoic3lzdGVtOnNlcnZpY2VhY2NvdW50OmdhcmRlbjpkZWZhdWx0In0.-4rSuvvj5BStN6DwnmLAaRVbgpl5iCn2hG0pcqx0NPw",
    },
  ],
  Array [
    Object {
      ":authority": "gardener-apiserver.gardener",
      ":method": "get",
      ":path": "/version",
      ":scheme": "https",
    },
  ],
]
`;

exports[`api info should return information without version 2`] = `
Object {
  "version": "1.47.0",
}
`;
